### PR TITLE
Error collection & refactoring in config parser & lark parser

### PR DESCRIPTION
**Pull request by yngve-sk**
**Issue**
Resolves https://github.com/equinor/ert/issues/5099

**Approach**
Uses list passed as parameters to catch errors from parsing, while still raising it at the end as a ConfigValidationError

## Pre review checklist

- [ ] Added appropriate release note label
- [ ] PR title captures the intent of the changes, and is fitting for release notes.
- [ ] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [ ] Updated documentation
- [ ] Ensured new behaviour is tested

Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).

### CODE REVIEW

File: src/ert/_c_wrappers/enkf/ert_config.py [hunk 1]
```diff
@@ -252,13 +298,17 @@ def read_user_config(
...
+        cls, config_path, config_dict, collected_errors: List[ErrorInfo]
+    ):
+        for queue_name, option_name, *values in config_dict.get("QUEUE_OPTION", []):
             if option_name == "MAX_RUNNING" and int(*values) < 0:
-                raise ConfigValidationError(
-                    errors=[
-                        f"QUEUE_OPTION MAX_RUNNING is negative: {str(*values)!r}",
-                    ],
+                collected_errors.append(
+                    ErrorInfo(
```
Comment (mortalisk) [inline src/ert/_c_wrappers/enkf/ert_config.py:307]: perhaps we can have an extra parameter to ErrorInfo, to which we feed potential FileContextTokens, and if they are tokens, ErrorInfo can copy all the relevant info from those tokens. If they are not, which will be the case in tests and in everests case, which do not use the parser, it can simply give general errors without filename, line, etc.

File: src/ert/_c_wrappers/enkf/ert_config.py [hunk 2]
```diff
@@ -70,50 +74,94 @@ class ErtConfig:
...
+        if collected_errors is None:
+            collected_errors = []
+            do_raise_errors = True
+
         substitution_list = SubstitutionList.from_dict(config_dict=config_dict)
         config_dir = substitution_list.get("<CONFIG_PATH>", "")
         config_file = substitution_list.get("<CONFIG_FILE>", "no_config")
         config_file_path = os.path.join(config_dir, config_file)
 
-        ErtConfig._validate_dict(config_dict, config_file)
```
Comment (mortalisk) [inline src/ert/_c_wrappers/enkf/ert_config.py:100]: We probably now desire to run as muc of the code as possible. Previously the code has been structured to stop early. I therefore propose that we ditch this _validate_dict, and let each part of the code that needs to detect error conditions which implies it cannot run, to detect and give an error in the list about that, and then simply skip calling that piece of the code, instead of skipping everything like before.

File: src/ert/_c_wrappers/config/config_parser.py [hunk 3]
```diff
@@ -9,6 +10,60 @@
...
+        ):
+            token = originates_from_keyword.keyword_token
+        elif originates_from is not None and hasattr(originates_from, "token"):
+            token = originates_from.token
+        elif originates_from_these is not None:
+            tokens = [x.token for x in originates_from_these if hasattr(x, "token")]
+            # Merge the token positions etc
+
+            if len(tokens) > 0:
+                from ert._c_wrappers.enkf.lark_parser_common import FileContextToken
```
Comment (DanSava) [inline src/ert/_c_wrappers/config/config_parser.py:53]: This should be a top level import

File: src/ert/_c_wrappers/enkf/ert_config.py [hunk 4]
```diff
@@ -69,58 +75,113 @@ class ErtConfig:
...
-            if self.user_config_file
-            else os.getcwd()
-        )
-
     @classmethod
     def from_file(
-        cls, user_config_file, use_new_parser: bool = USE_NEW_PARSER_BY_DEFAULT
+        cls,
+        user_config_file,
+        collected_errors: List[ErrorInfo] = None,
```
Comment (eivindjahren) [inline src/ert/_c_wrappers/enkf/ert_config.py:82]: I don't think it is a good idea to add this additional parameter. It seems to only be used in a test.

File: src/ert/_c_wrappers/enkf/ert_config.py [hunk 5]
```diff
@@ -69,58 +75,113 @@ class ErtConfig:
...
+        self.config_path = (
+            os.path.dirname(os.path.abspath(self.user_config_file))
+            if self.user_config_file
+            else os.getcwd()
+        )
 
     @classmethod
-    def from_dict(cls, config_dict) -> "ErtConfig":
+    def from_dict(
+        cls, config_dict, collected_errors: Optional[List[ErrorInfo]] = None
```
Comment (eivindjahren) [inline src/ert/_c_wrappers/enkf/ert_config.py:123]: Same with the addtional collected_errors parameter here. It doesn't seem to be used for anything.

File: src/ert/_c_wrappers/enkf/ert_config.py [hunk 6]
```diff
@@ -69,58 +75,113 @@ class ErtConfig:
...
+    ) -> "ErtConfig":
+        do_raise_errors = False
+        if collected_errors is None:
+            collected_errors = []
+            do_raise_errors = True
```
Comment (eivindjahren) [inline src/ert/_c_wrappers/enkf/ert_config.py:128]: This logic is fairly hard to follow. It would seem to me that we should always raise errors when there are some. in this function. Also, relying on the function not raising under certain conditions is going to be ahrd to manage. It would make more sense to always raise at the end, with the option of other parts of the code raising early if there are problems from which we cannot  continue.

File: src/ert/_c_wrappers/enkf/ert_config.py (hunk 4 above)
Comment (mortalisk) [reply src/ert/_c_wrappers/enkf/ert_config.py:82]: When implementing the linter, this will be used I guess?

File: src/ert/_c_wrappers/enkf/queue_config.py [hunk 7]
```diff
@@ -19,7 +19,7 @@ class QueueConfig:
     )
 
     @classmethod
-    def from_dict(cls, config_dict) -> QueueConfig:
+    def from_dict(cls, config_dict, collected_errors: List[ErrorInfo]) -> QueueConfig:
```
Comment (eivindjahren) [inline src/ert/_c_wrappers/enkf/queue_config.py:22]: The interface for this function becomes very difficult to follow, except in the context of `ErtConfig.from_file`. I suggest that it should raise ConfigValidationError that `ErtConfig.from_file` catches and then appends to its list of collected errors.

File: src/ert/parsing/config_errors.py [hunk 8]
```diff
@@ -17,3 +19,14 @@ def __init__(self, errors: str, config_file: Optional[str] = None) -> None:
             if self.config_file
             else f"{self.errors}"
         )
+
+    @classmethod
+    def raise_from_collected(cls, collected_errors: List[ErrorInfo]):
+        if len(collected_errors) > 0:
```
Comment (eivindjahren) [inline src/ert/parsing/config_errors.py:25]: For the GUI we need the list of errors as separate boxes, so I suggest going back to the approach of keeping the list of errors as a field.

File: src/ert/_c_wrappers/enkf/queue_config.py [hunk 9]
```diff
@@ -40,7 +40,19 @@ def from_dict(cls, config_dict) -> QueueConfig:
         max_submit = config_dict.get("MAX_SUBMIT", 2)
         queue_options = defaultdict(list)
         for driver, option_name, *values in config_dict.get("QUEUE_OPTION", []):
-            queue_driver_type = QueueDriverEnum.from_string(driver + "_DRIVER")
+            try:
+                queue_driver_type = QueueDriverEnum.from_string(driver + "_DRIVER")
+            except ValueError as err:
+                collected_errors.append(
+                    ErrorInfo(
```
Comment (eivindjahren) [inline src/ert/_c_wrappers/enkf/queue_config.py:47]: This is checked in the validation step I believe.
